File: android/pouchprov-core/src/main/kotlin/io/golioth/pouchprov/pouch/Pouch.kt
// Copyright (c) 2026 Jonathan Beri
// SPDX-License-Identifier: Apache-2.0

package io.golioth.pouchprov.pouch

import io.golioth.pouchprov.cbor.Cbor
import java.io.ByteArrayOutputStream

/**
 * Pouch wire framing: header, blocks and entries.
 *
 * Formats mirror the pouch reference implementation (github.com/golioth/pouch:
 * src/header.cddl, src/block.c, src/entry.c) and the Python `pouchlink.pouch`.
 *
 * Block:
 *   0..1  size (be16, excluding the size field itself, including the id byte)
 *   2     id byte: low 5 bits stream id (0 = entry block), 0x40 = first, 0x80 = last
 *   3..   payload
 *
 * Entry (within an entry block):
 *   0..1  data_len (be16)
 *   2..3  content_type (be16, IANA CoAP content format)
 *   4     path_len
 *   5..   path, then data
 */
object Pouch {
    const val POUCH_VERSION = 1

    const val CONTENT_TYPE_OCTET_STREAM = 42
    const val CONTENT_TYPE_JSON = 50
    const val CONTENT_TYPE_CBOR = 60

    const val BLOCK_ID_ENTRY = 0x00
    const val BLOCK_ID_MASK = 0x1F
    const val BLOCK_FIRST = 0x40
    const val BLOCK_LAST = 0x80
    const val BLOCK_HEADER_SIZE = 3

    const val ENTRY_HEADER_OVERHEAD = 5

    const val ENCRYPTION_NONE = 0
    const val ENCRYPTION_SAEAD = 1

    const val ROLE_DEVICE = 0
    const val ROLE_SERVER = 1

    const val ALG_CHACHA20_POLY1305 = 1
    const val ALG_AES_GCM = 2

    const val DEFAULT_BLOCK_SIZE = 512

    class DecodeException(message: String) : Exception(message)

    /**
     * Build a plaintext (encryption-none) pouch carrying [entries].
     *
     * Entries are packed into entry blocks of at most [blockSize] payload bytes;
     * an entry never spans blocks (pouch src/entry.c contract).
     */
    fun buildEntryPouch(deviceId: String, entries: List<Entry>, blockSize: Int = DEFAULT_BLOCK_SIZE): ByteArray {
        val blocks = ArrayList<Block>()
        var current = ByteArrayOutputStream()
        for (entry in entries) {
            val encoded = entry.encode()
            if (encoded.size > blockSize) {
                throw IllegalArgumentException("entry for ${entry.path} exceeds block size $blockSize")
            }
            if (current.size() > 0 && current.size() + encoded.size > blockSize) {
                blocks.add(Block(current.toByteArray(), first = blocks.isEmpty(), last = false))
                current = ByteArrayOutputStream()
            }
            current.write(encoded)
        }
        blocks.add(Block(current.toByteArray(), first = blocks.isEmpty(), last = true))

        val out = ByteArrayOutputStream()
        out.write(PouchHeader.plaintext(deviceId).encode())
        for (block in blocks) out.write(block.encode())
        return out.toByteArray()
    }

    /** Parse entries out of concatenated entry-block payloads. */
    fun parseEntryBlocks(payloads: List<ByteArray>): List<Entry> {
        val data = concat(payloads)
        val entries = ArrayList<Entry>()
        var pos = 0
        while (pos < data.size) {
            if (data.size - pos < ENTRY_HEADER_OVERHEAD) throw DecodeException("truncated entry header")
            val dataLen = be16(data, pos)
            val contentType = be16(data, pos + 2)
            val pathLen = data[pos + 4].toInt() and 0xFF
            pos += ENTRY_HEADER_OVERHEAD
            if (data.size - pos < pathLen + dataLen) throw DecodeException("truncated entry")
            val path = String(data, pos, pathLen, Charsets.UTF_8)
            pos += pathLen
            val payload = data.copyOfRange(pos, pos + dataLen)
            pos += dataLen
            entries.add(Entry(path, contentType, payload))
        }
        return entries
    }

    /**
     * Parse a plaintext pouch into its header and entries. saead pouches must
     * have their blocks decrypted first (out of scope for the plaintext path).
     */
    fun parsePouch(data: ByteArray): Pair<PouchHeader, List<Entry>> {
        val (header, consumed) = PouchHeader.decode(data)
        val rawBlocks = splitBlocks(data.copyOfRange(consumed, data.size))
        if (header.encryption != ENCRYPTION_NONE) throw DecodeException("encrypted pouch passed to parsePouch")
        val payloads = rawBlocks.filter { it.streamId == BLOCK_ID_ENTRY }.map { it.payload }
        return header to parseEntryBlocks(payloads)
    }

    data class RawBlock(val streamId: Int, val first: Boolean, val last: Boolean, val payload: ByteArray)

    /** Split the block section into raw blocks. */
    fun splitBlocks(data: ByteArray): List<RawBlock> {
        val blocks = ArrayList<RawBlock>()
        var pos = 0
        while (pos < data.size) {
            if (data.size - pos < BLOCK_HEADER_SIZE) throw DecodeException("truncated block header")
            val size = be16(data, pos)
            val idByte = data[pos + 2].toInt() and 0xFF
            val payloadLen = size - 1 // size includes the id byte
            pos += BLOCK_HEADER_SIZE
            if (payloadLen < 0 || data.size - pos < payloadLen) throw DecodeException("truncated block payload")
            blocks.add(
                RawBlock(
                    idByte and BLOCK_ID_MASK,
                    idByte and BLOCK_FIRST != 0,
                    idByte and BLOCK_LAST != 0,
                    data.copyOfRange(pos, pos + payloadLen),
                ),
            )
            pos += payloadLen
        }
        return blocks
    }

    private fun be16(data: ByteArray, off: Int): Int =
        ((data[off].toInt() and 0xFF) shl 8) or (data[off + 1].toInt() and 0xFF)

    private fun concat(chunks: List<ByteArray>): ByteArray {
        val out = ByteArrayOutputStream()
        for (c in chunks) out.write(c)
        return out.toByteArray()
    }
}

/** A single pouch entry: a message on a reserved `.prov/` path. */
data class Entry(val path: String, val contentType: Int, val data: ByteArray) {
    fun encode(): ByteArray {
        val pathBytes = path.toByteArray(Charsets.UTF_8)
        require(pathBytes.size <= 255) { "path too long" }
        val out = ByteArrayOutputStream()
        out.write((data.size ushr 8) and 0xFF)
        out.write(data.size and 0xFF)
        out.write((contentType ushr 8) and 0xFF)
        out.write(contentType and 0xFF)
        out.write(pathBytes.size and 0xFF)
        out.write(pathBytes)
        out.write(data)
        return out.toByteArray()
    }

    override fun equals(other: Any?): Boolean =
        other is Entry && path == other.path && contentType == other.contentType && data.contentEquals(other.data)

    override fun hashCode(): Int = (path.hashCode() * 31 + contentType) * 31 + data.contentHashCode()
}

/** A pouch block: an id byte plus a payload. */
data class Block(
    val payload: ByteArray,
    val streamId: Int = Pouch.BLOCK_ID_ENTRY,
    val first: Boolean = true,
    val last: Boolean = true,
) {
    fun encode(): ByteArray {
        var idByte = streamId and Pouch.BLOCK_ID_MASK
        if (first) idByte = idByte or Pouch.BLOCK_FIRST
        if (last) idByte = idByte or Pouch.BLOCK_LAST
        // size excludes the 2-byte size field but includes the id byte
        val size = payload.size + 1
        val out = ByteArrayOutputStream()
        out.write((size ushr 8) and 0xFF)
        out.write(size and 0xFF)
        out.write(idByte)
        out.write(payload)
        return out.toByteArray()
    }

    override fun equals(other: Any?): Boolean =
        other is Block && streamId == other.streamId && first == other.first &&
            last == other.last && payload.contentEquals(other.payload)

    override fun hashCode(): Int =
        ((payload.contentHashCode() * 31 + streamId) * 31 + first.hashCode()) * 31 + last.hashCode()
}

/**
 * Pouch header per src/header.cddl.
 *
 * Plaintext: `[1, [0, device_id]]`
 * saead:     `[1, [1, session_info, pouch_id]]` (decoded for detection; the
 *            encrypted payload path is a follow-up — see the plan's crypto seam).
 */
data class PouchHeader(
    val encryption: Int = Pouch.ENCRYPTION_NONE,
    val deviceId: String? = null,
    val sessionInfo: List<Any?>? = null,
    val pouchId: Long = 0,
) {
    fun encode(): ByteArray {
        val info: List<Any?> = if (encryption == Pouch.ENCRYPTION_NONE) {
            requireNotNull(deviceId) { "plaintext header requires device_id" }
            listOf(Pouch.ENCRYPTION_NONE.toLong(), deviceId)
        } else {
            requireNotNull(sessionInfo) { "saead header requires session info" }
            listOf(Pouch.ENCRYPTION_SAEAD.toLong(), sessionInfo, pouchId)
        }
        return Cbor.encode(listOf(Pouch.POUCH_VERSION.toLong(), info))
    }

    companion object {
        fun plaintext(deviceId: String) = PouchHeader(Pouch.ENCRYPTION_NONE, deviceId = deviceId)

        /** Decode a header from the start of [data]; returns it and bytes consumed. */
        fun decode(data: ByteArray): Pair<PouchHeader, Int> {
            val (obj, consumed) = try {
                Cbor.decodeWithLength(data)
            } catch (e: Exception) {
                throw Pouch.DecodeException("bad pouch header: ${e.message}")
            }
            if (obj !is List<*> || obj.size != 2) throw Pouch.DecodeException("pouch header is not a 2-element array")
            val version = (obj[0] as? Long)?.toInt()
            if (version != Pouch.POUCH_VERSION) throw Pouch.DecodeException("unsupported pouch version $version")
            val info = obj[1] as? List<*> ?: throw Pouch.DecodeException("bad encryption info")
            if (info.isEmpty()) throw Pouch.DecodeException("bad encryption info")
            return when ((info[0] as? Long)?.toInt()) {
                Pouch.ENCRYPTION_NONE -> PouchHeader(Pouch.ENCRYPTION_NONE, deviceId = info[1] as String?) to consumed
                Pouch.ENCRYPTION_SAEAD -> PouchHeader(
                    Pouch.ENCRYPTION_SAEAD,
                    sessionInfo = @Suppress("UNCHECKED_CAST") (info[1] as List<Any?>),
                    pouchId = (info[2] as? Long) ?: 0L,
                ) to consumed
                else -> throw Pouch.DecodeException("unknown encryption type ${info[0]}")
            }
        }
    }
}

File: ios/PouchProv/Sources/PouchProvCore/Pouch.swift
// Copyright (c) 2026 Jonathan Beri
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// Pouch wire framing: header, blocks and entries.
///
/// Formats mirror the pouch reference implementation (github.com/golioth/pouch:
/// src/header.cddl, src/block.c, src/entry.c) and the Kotlin/Python ports.
///
/// Block:
///   0..1  size (be16, excluding the size field itself, including the id byte)
///   2     id byte: low 5 bits stream id (0 = entry block), 0x40 = first, 0x80 = last
///   3..   payload
///
/// Entry (within an entry block):
///   0..1  data_len (be16)
///   2..3  content_type (be16, IANA CoAP content format)
///   4     path_len
///   5..   path, then data
public enum Pouch {
    public static let pouchVersion = 1

    public static let contentTypeOctetStream = 42
    public static let contentTypeJson = 50
    public static let contentTypeCbor = 60

    public static let blockIdEntry = 0x00
    public static let blockIdMask = 0x1F
    public static let blockFirst = 0x40
    public static let blockLast = 0x80
    public static let blockHeaderSize = 3

    public static let entryHeaderOverhead = 5

    public static let encryptionNone = 0
    public static let encryptionSaead = 1

    public static let roleDevice = 0
    public static let roleServer = 1

    public static let algChacha20Poly1305 = 1
    public static let algAesGcm = 2

    public static let defaultBlockSize = 512

    public struct DecodeException: Error, CustomStringConvertible {
        public let message: String

        public init(_ message: String) {
            self.message = message
        }

        public var description: String { message }
    }

    public struct EncodeException: Error, CustomStringConvertible {
        public let message: String

        public init(_ message: String) {
            self.message = message
        }

        public var description: String { message }
    }

    /// Build a plaintext (encryption-none) pouch carrying `entries`.
    ///
    /// Entries are packed into entry blocks of at most `blockSize` payload bytes;
    /// an entry never spans blocks (pouch src/entry.c contract).
    public static func buildEntryPouch(
        deviceId: String,
        entries: [Entry],
        blockSize: Int = defaultBlockSize
    ) throws -> Data {
        var blocks: [Block] = []
        var current = Data()
        for entry in entries {
            let encoded = entry.encode()
            if encoded.count > blockSize {
                throw EncodeException("entry for \(entry.path) exceeds block size \(blockSize)")
            }
            if !current.isEmpty && current.count + encoded.count > blockSize {
                blocks.append(Block(payload: current, first: blocks.isEmpty, last: false))
                current = Data()
            }
            current.append(encoded)
        }
        blocks.append(Block(payload: current, first: blocks.isEmpty, last: true))

        var out = PouchHeader.plaintext(deviceId: deviceId).encode()
        for block in blocks { out.append(block.encode()) }
        return out
    }

    /// Parse entries out of concatenated entry-block payloads.
    public static func parseEntryBlocks(_ payloads: [Data]) throws -> [Entry] {
        var data = Data()
        for payload in payloads { data.append(payload) }
        let bytes = [UInt8](data)
        var entries: [Entry] = []
        var pos = 0
        while pos < bytes.count {
            guard bytes.count - pos >= entryHeaderOverhead else {
                throw DecodeException("truncated entry header")
            }
            let dataLen = be16(bytes, pos)
            let contentType = be16(bytes, pos + 2)
            let pathLen = Int(bytes[pos + 4])
            pos += entryHeaderOverhead
            guard bytes.count - pos >= pathLen + dataLen else {
                throw DecodeException("truncated entry")
            }
            guard let path = String(bytes: bytes[pos..<(pos + pathLen)], encoding: .utf8) else {
                throw DecodeException("bad entry path")
            }
            pos += pathLen
            let payload = Data(bytes[pos..<(pos + dataLen)])
            pos += dataLen
            entries.append(Entry(path: path, contentType: contentType, data: payload))
        }
        return entries
    }

    /// Parse a plaintext pouch into its header and entries. saead pouches must
    /// have their blocks decrypted first (out of scope for the plaintext path).
    public static func parsePouch(_ data: Data) throws -> (header: PouchHeader, entries: [Entry]) {
        let (header, consumed) = try PouchHeader.decode(data)
        let rawBlocks = try splitBlocks(data.dropFirst(consumed))
        guard header.encryption == encryptionNone else {
            throw DecodeException("encrypted pouch passed to parsePouch")
        }
        let payloads = rawBlocks.filter { $0.streamId == blockIdEntry }.map(\.payload)
        return (header, try parseEntryBlocks(payloads))
    }

    public struct RawBlock: Equatable, Sendable {
        public let streamId: Int
        public let first: Bool
        public let last: Bool
        public let payload: Data
    }

    /// Split the block section into raw blocks.
    public static func splitBlocks(_ data: Data) throws -> [RawBlock] {
        let bytes = [UInt8](data)
        var blocks: [RawBlock] = []
        var pos = 0
        while pos < bytes.count {
            guard bytes.count - pos >= blockHeaderSize else {
                throw DecodeException("truncated block header")
            }
            let size = be16(bytes, pos)
            let idByte = Int(bytes[pos + 2])
            let payloadLen = size - 1 // size includes the id byte
            pos += blockHeaderSize
            guard payloadLen >= 0, bytes.count - pos >= payloadLen else {
                throw DecodeException("truncated block payload")
            }
            blocks.append(RawBlock(
                streamId: idByte & blockIdMask,
                first: idByte & blockFirst != 0,
                last: idByte & blockLast != 0,
                payload: Data(bytes[pos..<(pos + payloadLen)])
            ))
            pos += payloadLen
        }
        return blocks
    }

    private static func be16(_ bytes: [UInt8], _ offset: Int) -> Int {
        (Int(bytes[offset]) << 8) | Int(bytes[offset + 1])
    }
}

/// A single pouch entry: a message on a reserved `.prov/` path.
public struct Entry: Equatable, Sendable {
    public let path: String
    public let contentType: Int
    public let data: Data

    public init(path: String, contentType: Int = Pouch.contentTypeCbor, data: Data) {
        self.path = path
        self.contentType = contentType
        self.data = data
    }

    public func encode() -> Data {
        let pathBytes = Data(path.utf8)
        precondition(pathBytes.count <= 255, "path too long")
        var out = Data()
        out.append(UInt8((data.count >> 8) & 0xFF))
        out.append(UInt8(data.count & 0xFF))
        out.append(UInt8((contentType >> 8) & 0xFF))
        out.append(UInt8(contentType & 0xFF))
        out.append(UInt8(pathBytes.count))
        out.append(pathBytes)
        out.append(data)
        return out
    }
}

/// A pouch block: an id byte plus a payload.
public struct Block: Equatable, Sendable {
    public let payload: Data
    public let streamId: Int
    public let first: Bool
    public let last: Bool

    public init(payload: Data, streamId: Int = Pouch.blockIdEntry, first: Bool = true, last: Bool = true) {
        self.payload = payload
        self.streamId = streamId
        self.first = first
        self.last = last
    }

    public func encode() -> Data {
        var idByte = streamId & Pouch.blockIdMask
        if first { idByte |= Pouch.blockFirst }
        if last { idByte |= Pouch.blockLast }
        // size excludes the 2-byte size field but includes the id byte
        let size = payload.count + 1
        var out = Data()
        out.append(UInt8((size >> 8) & 0xFF))
        out.append(UInt8(size & 0xFF))
        out.append(UInt8(idByte))
        out.append(payload)
        return out
    }
}

/// Pouch header per src/header.cddl.
///
/// Plaintext: `[1, [0, device_id]]`
/// saead:     `[1, [1, session_info, pouch_id]]` (decoded for detection; the
///            encrypted payload path is a follow-up — see the `SessionCrypto` seam).
public struct PouchHeader: Equatable, Sendable {
    public let encryption: Int
    public let deviceId: String?
    public let sessionInfo: [CborValue]?
    public let pouchId: Int64

    public init(
        encryption: Int = Pouch.encryptionNone,
        deviceId: String? = nil,
        sessionInfo: [CborValue]? = nil,
        pouchId: Int64 = 0
    ) {
        self.encryption = encryption
        self.deviceId = deviceId
        self.sessionInfo = sessionInfo
        self.pouchId = pouchId
    }

    public static func plaintext(deviceId: String) -> PouchHeader {
        PouchHeader(encryption: Pouch.encryptionNone, deviceId: deviceId)
    }

    public func encode() -> Data {
        let info: CborValue
        if encryption == Pouch.encryptionNone {
            guard let deviceId else { preconditionFailure("plaintext header requires device_id") }
            info = .array([.int(Int64(Pouch.encryptionNone)), .text(deviceId)])
        } else {
            guard let sessionInfo else { preconditionFailure("saead header requires session info") }
            info = .array([.int(Int64(Pouch.encryptionSaead)), .array(sessionInfo), .int(pouchId)])
        }
        return Cbor.encode(.array([.int(Int64(Pouch.pouchVersion)), info]))
    }

    /// Decode a header from the start of `data`; returns it and bytes consumed.
    public static func decode(_ data: Data) throws -> (header: PouchHeader, consumed: Int) {
        let obj: CborValue
        let consumed: Int
        do {
            (obj, consumed) = try Cbor.decodeWithLength(data)
        } catch {
            throw Pouch.DecodeException("bad pouch header: \(error)")
        }
        guard let outer = obj.arrayValue, outer.count == 2 else {
            throw Pouch.DecodeException("pouch header is not a 2-element array")
        }
        guard let version = outer[0].intValue, version == Pouch.pouchVersion else {
            throw Pouch.DecodeException("unsupported pouch version \(outer[0])")
        }
        guard let info = outer[1].arrayValue, !info.isEmpty else {
            throw Pouch.DecodeException("bad encryption info")
        }
        switch info[0].intValue {
        case Pouch.encryptionNone:
            guard info.count > 1 else { throw Pouch.DecodeException("bad encryption info") }
            return (PouchHeader(encryption: Pouch.encryptionNone, deviceId: info[1].textValue), consumed)
        case Pouch.encryptionSaead:
            guard info.count > 2, let sessionInfo = info[1].arrayValue else {
                throw Pouch.DecodeException("bad encryption info")
            }
            return (
                PouchHeader(
                    encryption: Pouch.encryptionSaead,
                    sessionInfo: sessionInfo,
                    pouchId: info[2].int64Value ?? 0
                ),
                consumed
            )
        default:
            throw Pouch.DecodeException("unknown encryption type \(info[0])")
        }
    }
}
